# Hate Crimes Analysis - Book2 (Rough Draft)

## Data Cleaning

In the last notebook, I merged datasets from data.austintexas.gov on the years for which Austin began recording alleged reported hate crimes (2017 - Present). 

Now, I'm going to begin the process of cleaning up the newly concatenated dataset to get it ready for analysis. 

In [2]:
import pandas as pd
import numpy as np

In [3]:
# Loading the concatenated dataset from the last notebook
df = pd.read_csv(r"C:\Users\Robert\OneDrive\Desktop\aus_final.csv", 
                 keep_default_na=False, 
                 na_values=[""])
display(df.describe())
display(df.head())
display(df.tail())
display(df.info())

Unnamed: 0  number_of_offenders_over  number_of_offenders_over_18  \
count   53.000000                 17.000000                    36.000000   
mean     7.226415                  1.117647                     1.000000   
std      5.138997                  1.111438                     0.676123   
min      0.000000                  0.000000                     0.000000   
25%      3.000000                  0.000000                     1.000000   
50%      7.000000                  1.000000                     1.000000   
75%     11.000000                  1.000000                     1.000000   
max     18.000000                  4.000000                     4.000000   

       number_of_offenders_under  number_of_offenders_under_18  \
count                  17.000000                     36.000000   
mean                    0.176471                      0.111111   
std                     0.727607                      0.398410   
min                     0.000000                      0.000000   
25%                     0.000000                      0.000000   
50%                     0.000000                      0.000000   
75%                     0.000000                      0.000000   
max                     3.000000                      2.000000   

       number_of_victims_over_18  number_of_victims_under_18  \
count                  53.000000                   17.000000   
mean                    0.905660                    0.058824   
std                     0.404976                    0.242536   
min                     0.000000                    0.000000   
25%                     1.000000                    0.000000   
50%                     1.000000                    0.000000   
75%                     1.000000                    0.000000   
max                     2.000000                    1.000000   

       number_of_vitims_under_18  
count                  36.000000  
mean                    0.055556  
std                     0.232311  
min                     0.000000  
25%                     0.000000  
50%                     0.000000  
75%                     0.000000  
max                     1.000000

Unnamed: 0                            bias date_of_incident  \
0           0  Anti-Black or African American              NaN   
1           1                      Anti-White              NaN   
2           2                     Anti-Jewish              NaN   
3           3                     Anti-Jewish              NaN   
4           4                 Anti-Gay (Male)              NaN   

  date_of_incident_day_of_week day_of_week incident_number     month notes  \
0               01/01/2017/Sun         NaN     2017-241137   January   NaN   
1               02/01/2017/Wed         NaN     2017-580344  February   NaN   
2              03/21/2017/Tues         NaN     2017-800291     March   NaN   
3               04/12/2017/Wed         NaN    2017-1021534     April   NaN   
4               05/15/2017/Mon         NaN    2017-1351550       May   NaN   

   number_of_offenders_over  number_of_offenders_over_18  ...  \
0                       NaN                          1.0  ...   
1                       NaN                          1.0  ...   
2                       NaN                          0.0  ...   
3                       NaN                          0.0  ...   
4                       NaN                          2.0  ...   

   number_of_victims_over_18  number_of_victims_under_18  \
0                          1                         NaN   
1                          1                         NaN   
2                          0                         NaN   
3                          0                         NaN   
4                          0                         NaN   

   number_of_vitims_under_18             offense  \
0                        0.0  Aggravated Assault   
1                        0.0  Aggravated Assault   
2                        0.0         Destruction   
3                        0.0      Simple Assault   
4                        1.0      Simple Assault   

                     offense_location offense_s race_ethnic_of_offender_s  \
0                     Park/Playground       NaN                       NaN   
1  Highway/Road/Alley/Street/Sidewalk       NaN                       NaN   
2  Highway/Road/Alley/Street/Sidewalk       NaN                       NaN   
3              Air/Bus/Train Terminal       NaN                       NaN   
4                      Residence/Home       NaN                       NaN   

  race_ethnicity_of_offenders              race_or_ethnic_of_offender  \
0                         NaN                      White/Not Hispanic   
1                         NaN  Black or African American/Not Hispanic   
2                         NaN                                 Unknown   
3                         NaN                           White/Unknown   
4                         NaN                      White/Not Hispanic   

  victim_type  
0  Individual  
1  Individual  
2       Other  
3  Individual  
4  Individual  

[5 rows x 22 columns]

Unnamed: 0                               bias         date_of_incident  \
48           0                         Anti-Black  2020-03-01T00:00:00.000   
49           1  Anti-Gay (Male); Anti-Transgender  2020-03-08T00:00:00.000   
50           2       Anti-Gay (Male); Anti-Jewish  2020-03-22T00:00:00.000   
51           3                      Anti-Buddhist  2020-04-20T00:00:00.000   
52           4  Anti-Gay (Male); Anti-Transgender  2020-04-29T00:00:00.000   

   date_of_incident_day_of_week day_of_week incident_number  month notes  \
48                          NaN      Sunday     2020-602085  March   NaN   
49                          NaN      Sunday     2020-680226  March   NaN   
50                          NaN      Sunday    2020-5011788  March   NaN   
51                          NaN      Monday    2020-5015689  April   NaN   
52                          NaN   Wednesday    2020-5016804  April   NaN   

    number_of_offenders_over  number_of_offenders_over_18  ...  \
48                       1.0                          NaN  ...   
49                       2.0                          NaN  ...   
50                       0.0                          NaN  ...   
51                       0.0                          NaN  ...   
52                       1.0                          NaN  ...   

    number_of_victims_over_18  number_of_victims_under_18  \
48                          1                         0.0   
49                          1                         0.0   
50                          1                         0.0   
51                          1                         0.0   
52                          1                         0.0   

    number_of_vitims_under_18  offense                offense_location  \
48                        NaN      NaN                  Residence/Home   
49                        NaN      NaN                     Parking Lot   
50                        NaN      NaN                  Residence/Home   
51                        NaN      NaN  Church/Synagogue/Temple/Mosque   
52                        NaN      NaN       Department/Discount Store   

            offense_s race_ethnic_of_offender_s race_ethnicity_of_offenders  \
48  Criminal Mischief                       NaN          White/Non-Hispanic   
49            Assault                       NaN              White/Hispanic   
50  Criminal Mischief                       NaN                     Unknown   
51  Criminal Mischief                       NaN                     Unknown   
52  Assault by Threat                       NaN          Black/Non-Hispanic   

   race_or_ethnic_of_offender victim_type  
48                        NaN         NaN  
49                        NaN         NaN  
50                        NaN         NaN  
51                        NaN         NaN  
52                        NaN         NaN  

[5 rows x 22 columns]

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 53 entries, 0 to 52
Data columns (total 22 columns):
Unnamed: 0                      53 non-null int64
bias                            53 non-null object
date_of_incident                17 non-null object
date_of_incident_day_of_week    36 non-null object
day_of_week                     17 non-null object
incident_number                 53 non-null object
month                           53 non-null object
notes                           2 non-null object
number_of_offenders_over        17 non-null float64
number_of_offenders_over_18     36 non-null float64
number_of_offenders_under       17 non-null float64
number_of_offenders_under_18    36 non-null float64
number_of_victims_over_18       53 non-null int64
number_of_victims_under_18      17 non-null float64
number_of_vitims_under_18       36 non-null float64
offense                         17 non-null object
offense_location                53 non-null object
offense_s                  

None

In [4]:
# Checking for all null values
display(df.isnull())
display(df.notnull())
display(df.nunique())

Unnamed: 0   bias  date_of_incident  date_of_incident_day_of_week  \
0        False  False              True                         False   
1        False  False              True                         False   
2        False  False              True                         False   
3        False  False              True                         False   
4        False  False              True                         False   
5        False  False              True                         False   
6        False  False              True                         False   
7        False  False              True                         False   
8        False  False              True                         False   
9        False  False              True                         False   
10       False  False              True                         False   
11       False  False              True                         False   
12       False  False              True                         False   
13       False  False              True                         False   
14       False  False              True                         False   
15       False  False              True                         False   
16       False  False              True                         False   
17       False  False              True                         False   
18       False  False              True                         False   
19       False  False              True                         False   
20       False  False              True                         False   
21       False  False              True                         False   
22       False  False              True                         False   
23       False  False              True                         False   
24       False  False              True                         False   
25       False  False              True                         False   
26       False  False              True                         False   
27       False  False              True                         False   
28       False  False              True                         False   
29       False  False              True                         False   
30       False  False              True                         False   
31       False  False              True                         False   
32       False  False              True                         False   
33       False  False              True                         False   
34       False  False              True                         False   
35       False  False              True                         False   
36       False  False             False                          True   
37       False  False             False                          True   
38       False  False             False                          True   
39       False  False             False                          True   
40       False  False             False                          True   
41       False  False             False                          True   
42       False  False             False                          True   
43       False  False             False                          True   
44       False  False             False                          True   
45       False  False             False                          True   
46       False  False             False                          True   
47       False  False             False                          True   
48       False  False             False                          True   
49       False  False             False                          True   
50       False  False             False                          True   
51       False  False             False                          True   
52       False  False             False                          True   

    day_of_week  incident_number  month  notes  number_of_off

Unnamed: 0  bias  date_of_incident  date_of_incident_day_of_week  \
0         True  True             False                          True   
1         True  True             False                          True   
2         True  True             False                          True   
3         True  True             False                          True   
4         True  True             False                          True   
5         True  True             False                          True   
6         True  True             False                          True   
7         True  True             False                          True   
8         True  True             False                          True   
9         True  True             False                          True   
10        True  True             False                          True   
11        True  True             False                          True   
12        True  True             False                          True   
13        True  True             False                          True   
14        True  True             False                          True   
15        True  True             False                          True   
16        True  True             False                          True   
17        True  True             False                          True   
18        True  True             False                          True   
19        True  True             False                          True   
20        True  True             False                          True   
21        True  True             False                          True   
22        True  True             False                          True   
23        True  True             False                          True   
24        True  True             False                          True   
25        True  True             False                          True   
26        True  True             False                          True   
27        True  True             False                          True   
28        True  True             False                          True   
29        True  True             False                          True   
30        True  True             False                          True   
31        True  True             False                          True   
32        True  True             False                          True   
33        True  True             False                          True   
34        True  True             False                          True   
35        True  True             False                          True   
36        True  True              True                         False   
37        True  True              True                         False   
38        True  True              True                         False   
39        True  True              True                         False   
40        True  True              True                         False   
41        True  True              True                         False   
42        True  True              True                         False   
43        True  True              True                         False   
44        True  True              True                         False   
45        True  True              True                         False   
46        True  True              True                         False   
47        True  True              True                         False   
48        True  True              True                         False   
49        True  True              True                         False   
50        True  True              True                         False   
51        True  True              True                         False   
52        True  True              True                         False   

    day_of_week  incident_number  month  notes  number_of_offenders_over  \
0         False             True   True

Unnamed: 0                      19
bias                            18
date_of_incident                17
date_of_incident_day_of_week    36
day_of_week                      7
incident_number                 53
month                           15
notes                            2
number_of_offenders_over         5
number_of_offenders_over_18      4
number_of_offenders_under        2
number_of_offenders_under_18     3
number_of_victims_over_18        3
number_of_victims_under_18       2
number_of_vitims_under_18        2
offense                          5
offense_location                21
offense_s                        8
race_ethnic_of_offender_s        6
race_ethnicity_of_offenders      8
race_or_ethnic_of_offender       7
victim_type                      4
dtype: int64

Obviously, we have a lot of null values because the concatenation process created new tables of data since the different sets were not formatted alike (sloppy work on the curators/creators of the original sets, #notsorry). We'll need to take several steps to clean this mess up. 

    1. Merge the different victims & offenders columns, which contain the only numerical data in the set. (One 'number_of_victims_over_18' column is actually in-tact and holds no null values so we'll cut down our labor time by just deleting all the other like-named columns). 
    
    2. See what remaining columns can be converted into numerical data for better visualization.

In [5]:
# Focusing on the bias column again, let's see what it has to tell us. 
df['bias'].describe()

count                  53
unique                 18
top       Anti-Gay (Male)
freq                   10
Name: bias, dtype: object

##### Already without having officially begun the actual analysis phase, we can see that 'Anti-Gay (Male)' is the largest single motivational bias for reported hate crimes in Austin. 
##### That alone is 19% of all reported cases since the beginning of 2017 to date, and we have not even merged that bias into one category, with the other types of anti-lgbt biases!